### PR TITLE
[3.x] Prevent creating folders or files starting with a dot

### LINKED ISSUE #62497: A scene name can be empty (`.tscn`), leading to confusion in the list of open scenes and sometimes disappearing from the FileSystem dock
### Godot version

4.0.alpha10 (likely applies to 3.x too)

### System information

Fedora 36, GeForce GTX 1080 (NVIDIA 510.68.02)

### Issue description

Godot editor allows you to save a file with no name but an extension, such as `.tscn` or `.scn`. The result is that the scene will appear to have no name in the list of open scenes, which can look confusing. On top of that, the scene file will sometimes disappear and reappear in the FileSystem dock.

image

I suggest not allowing empty names for saved scenes in the editor – the lefthand side (without the extension) should be at least 1 character long.

### Steps to reproduce

- Create a new scene and add any node.
- Press <kbd>Ctrl + S</kbd> to save the scene, and change its name to `.tscn` or `.scn`.
- Notice how the scene will not have any name (as the file extension is trimmed), and it will intermittently disappear from the FileSystem dock.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by gotnospirit**
This PR is intended to fix #62497 on **3.x branch**.

It prevents the creation/renaming/duplication of a folder/scene/script/resource from the FileSystemDock while using a name with leading dot.
"_EditorDirDialog_" and "_EditorFileDialog_" were also modified to prevent creating such folder.

Also, as suggested, it prevents the creation of a scene like "a/my_scene.tscn". As a side-effect, this PR also avoid creating a scene as "res://my.tscn", which actually is rejected when you try to save the scene (_"res://res/ no longer exists! Please specify a new save location."_).

It introduce 2 new translations :
- "Could not use a name with a leading dot."
- "Could not create a scene for another folder.\nPlease select the target folder first."

I didn't modify "_scene\gui\FileDialog::_make_dir_confirm_" because I wasn't sure of possible side-effects

### CODE REVIEW

**Comment by SekoiaTree** [comment]
Erm, it looks like this is PRing to master? There's a few too many commits here :sweat_smile:

**Comment by gotnospirit** [comment]
Should I close this PR? or maybe there's an option I missed while using github?

**Comment by gotnospirit** [comment]
It worked, thanks!

**Review by Calinou — APPROVED** [review]
Looks good to me, but this needs to be rebased as code in for handling file/folder creation in `editor/filesystem_dock.cpp` has changed significantly. Also, a PR for `master` must be provided before this can be merged in 3.x.

@SekoiaTree Are you still available to write a `master` version of this PR once this one is rebased? Note that some cases were handled by https://github.com/godotengine/godot/pull/69801 already.

**Comment by SekoiaTree** [comment]
Yeah, sure, I can do that. I'll try to have it done next week.

**Comment by akien-mga** [comment]
Needs rebase and addressing the review comment. I didn't check, but maybe some code adjustment are needed to match what got merged in `master` in the end in #81075?

File: editor/filesystem_dock.cpp [hunk 1]
```diff
@@ -1457,6 +1466,9 @@ void FileSystemDock::_rename_operation_confirm() {
 	} else if (new_name.find("/") != -1 || new_name.find("\\") != -1 || new_name.find(":") != -1) {
 		EditorNode::get_singleton()->show_warning(TTR("Name contains invalid characters."));
 		return;
+	} else if (new_name.begins_with(".")) {
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:1469]:
```suggestion
	} else if (new_name[0] == '.') {
```

File: editor/filesystem_dock.cpp [hunk 2]
```diff
@@ -1519,6 +1531,9 @@ void FileSystemDock::_duplicate_operation_confirm() {
...
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:1534]:
```suggestion
	} else if (new_name[0] == '.') {
```

**Review by AThousandShips — COMMENTED** [review]
Should be the valid cases, can also do `string.length() != 0 && string[0] == '.'` for some cases where we don't know the length and it isn't something like `get_file` should still be more efficient but maybe less clear

**Comment by akien-mga** [comment]
@gotnospirit Are you available to address the review comments?

**Comment by gotnospirit** [comment]
yes, no problem

**Review by AThousandShips — APPROVED** [review]

**Comment by gotnospirit** [comment]
While making the changes I noticed what seems to be a bug
https://github.com/godotengine/godot/blob/4c4cb12e38dd862c1eedcb8ad17ca39ae9e3474e/editor/directory_create_dialog.cpp#L57-L61

I think `p_path` should be `path`

**Comment by SekoiaTree** [comment]
Yep, I noticed that as well when making the master version: https://github.com/godotengine/godot/pull/81075#discussion_r1316430537

**Comment by akien-mga** [comment]
I would suggest re-syncing the `3.x` implementation with #81075, which was initially based off this PR but evolved a lot under review. At least to the extent that the changes done there are compatible with the `3.x` codebase.

**Review by AThousandShips — APPROVED** [review]

**Pull request merged**